### ISSUE TITLE
Habit count/history feature proposal/discussion

### ISSUE THREAD

**Issue by jakechen** [HabitRPG/habitica#4544]
Been chatting with @Alys and in Aspiring Coders/Legends regarding a Habit Count/History feature that I will be personally working on. I would like to kickoff discussions on this feature.

It sounds like there's a good amount of demand for the ability to:
1. see how many times you completed a habit in a given day
2. see how many times you completed a habit over time (similar to value)
3. export the counts over time with the "export data" feature

I propose adding a "count" field to user.habits.history. This field will be modified when pluses or minuses are clicked. It will also be preened at the same time user.habits.history.value is preened. I plan on hopefully reusing most of that code.

Totally open to ideas here, I'm new to HabitRPG contributions so I'm sure there are intricacies that I don't know yet. Some things off the top of my head:
- Data cost - will adding a new field drastically impact cost?
- Functions modifying history - besides preenHistory, are there other backend processes that will need looking into?
- Other features - are there other feature requests that I could tie in here?

Please let me know if there's a better way to achieve the features above. Also let me know if this should be moved over to habitrpg-shared. Thanks in advance!
##

**Comment by deilann** [comment]
(this should definitely _not_ be moved into shared as we are working on
getting rid of shared)

**Comment by Alys** [comment]
Can you give us some information about what you have in mind for preening the count field? For example, I think you said elsewhere that you'd change the count to zero at the start of each day - is that correct?

**Comment by jakechen** [comment]
Sure thing. I think the resetting the count was Ryan's idea. My idea was to mirror how Values are currently done: if it's a new day then add a new date/value pair to history, else increment current date/value pair. This logic is shown here:

```
if th[th.length-1] and moment(th[th.length-1].date).isSame(new Date, 'day')
    th[th.length-1].value = task.value
else
    th.push {date: +new Date, value: task.value}
```

(https://github.com/HabitRPG/habitrpg-shared/blob/develop/script/index.coffee#L1078)

This means
1. daily preening isn't needed since every day will have a separate score
2. I can simply take advantage of this function to increment the score while the value is incremented

However, I know we have that other issue going on (#3079) to expand out the history so that each event is tracked instead of incrementing. That issue could impact this feature since the habit.history schema will change. It might make sense to do that one first before this one, but it could take a bit. This one on the other hand shouldn't be too bad?

**Comment by Alys** [comment]
I reckon ignore #3079 for now. We've certainly been ignoring it for a while. Don't let its existence hold you back from making a new feature.

If I understand correctly, your change would cause the user.habits.history object to look like this (new 'count' fields):

```
"history": [
                { 
                    "date": 1416307516281,
                    "value": 3.853133245658556,
                    "count": 4
                },
                {
                    "date": 1416358551417,
                    "value": 10.567273646958993,
                    "count": 12
                },
                {
                    "value": 11.330048651176607,
                    "date": 1416479141665,
                    "count": 0
                },
```

Is that right?

If @lefnire doesn't reply here with comments about data cost, we can track him down elsewhere. (He's the co-founder, chief programmer, chief database admin.)

**Comment by jakechen** [comment]
Yep, that's correct. In that case I'll skip past #3079 for now.

On Thursday, January 22, 2015, Alice Harris notifications@github.com
wrote:

> I reckon ignore #3079 https://github.com/HabitRPG/habitrpg/issues/3079
> for now. We've certainly been ignoring it for a while. Don't let its
> existence hold you back from making a new feature.
> 
> If I understand correctly, your change would cause the user.habits.history
> object to look like this (new 'count' fields):
> 
> "history": [
>                 {
>                     "date": 1416307516281,
>                     "value": 3.853133245658556,
>                     "count": 4
>                 },
>                 {
>                     "date": 1416358551417,
>                     "value": 10.567273646958993,
>                     "count": 12
>                 },
>                 {
>                     "value": 11.330048651176607,
>                     "date": 1416479141665,
>                     "count": 0
>                 },
> 
> Is that right?
> 
> If @lefnire https://github.com/lefnire doesn't reply here with comments
> about data cost, we can track him down elsewhere. (He's the co-founder,
> chief programmer, chief database admin.)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/issues/4544#issuecomment-71003941.

## 

Sent from a mobile device, please excuse any overeager autocorrects.

**Comment by paglias** [comment]
@jakechen @Alys 

A few random notes:
1. If we make this change then we'll have to calculate the value for each score on habits manually when there's more than one per day. I mean, if we click 'plus' on a habit twice in a single day the database would have:

```
                {
                    "value": 11.330048651176607,
                    "date": 1416479141665,
                    "count": 3
                },
```

To get the values (in graphs for example) of the single events on the habits we would have to calculate them every time.
1. Data cost should not go up substantially but of course you'll want too hear from @lefnire 
2. It would require a massive migration with probably some downtime...
3. "daily preening isn't needed since every day will have a separate score" I'm not sure but data preening is not only due to performances/data storage costs. It's possible we might want to keep it as a premium feature anyway. But maybe here I'm missing what you want to say

Let me know if any of the points above are not clear

**Comment by jakechen** [comment]
@paglias thanks for the notes.
1. Can you elaborate on "we'll have to calculate the value for each score on habits manually when there's more than one per day"?
...
1. Yup, holding on until I hear back.
2. I didn't think it would be too massive :( but you would know better.
3. Let me rephrase that. Daily preening as defined by "resetting the score every day" wouldn't be needed. That was one of the proposed solutions, but I believe it wouldn't be viable long-term. This current solution , adding a count field to habits.history, would theoretically use the same preening function as values i.e. preenHistory. I still have fully understand that function though.

**Comment by Alys** [comment]
@paglias The migration you spoke about - is that to add an estimated (or zero) `count` element to all existing history entries? If so, I think we should not bother and just make sure that all code that displays/exports counts gracefully ignores any entries that don't have a count value.

_"To get the values (in graphs for example) of the single events on the habits we would have to calculate them every time."_  We have to do that now, don't we? Single events aren't recorded (one of the reasons why a separate count field is needed). When a new habit click is done, the value for the previous habit click(s) for that day is overwritten, rather than a new value being added. https://github.com/HabitRPG/habitrpg-shared/blob/32ed4dace7b6d713e659e6f0b7ff7cea0160a75b/script/index.coffee#L1108-L1112

**Comment by taldin** [comment]
Here's an extra twist for you to think about:   
- if I click a '-' habit, I should get -1 point, -2 for the second bad habit today, and so on.
- This is because if I have a +/- habit, + should make a count in the positive direction, but flip to -1 if you subsequently click '-', and vice versa.

**Comment by paglias** [comment]
@jakechen 
1. @Alys is right, we don't store multiple events on the same day so ignore my previous comment
2. Since we don't store multiple events we don't need the migration either
3. Ok, got it

So, sorry for the confusion :)

**Comment by jakechen** [comment]
Perfect, sounds like my original plan should still get the job done :) just
waiting on the storage cost sign-off from the boss upstairs.

On Friday, January 23, 2015, Matteo Pagliazzi notifications@github.com
wrote:

> @jakechen https://github.com/jakechen
> 1. @Alys https://github.com/Alys is right, we don't store multiple
>    events on the same day so ignore my previous comment
> 2. Since we don't store multiple events we don't need the migration
>    either
> 3. Ok, got it
> 
> So, sorry for the confusion :)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/issues/4544#issuecomment-71258341.

## 

Sent from a mobile device, please excuse any overeager autocorrects.

**Comment by lefnire** [comment]
Sorry for the delay everyone. 
- Storage: Adding a single field to history entries is nuthin, :+1:. Especially since old entries will be preened automatically (should it be "pruned"?). 
- Preening: On that note, all you'd have to do in the preening process @jakechen is average `count` similar to how `value` is currently avg'd. Shouldn't be tough.
- #3079: Indeed, action history is a long-term goal. If you follow the discussion at https://github.com/HabitRPG/habitrpg-shared/pull/95, action history would solve more than a few issues (including a proper undo feature); however, that's so far out (at least, nobody's working on it that I know of) that I'd say roll with your implementation and we can deal with any migration ramifications if/when that time comes.

:+1: from my end.

**Comment by rickardnorlander** [comment]
If it is possible storage wise, I would like to be able to download a full, raw history. I.e. stuff like
[{"date": 1416307516281,item_id=34758,action="plus"}, {"date": 14163091452656,item_id=43219,action="dailyunchecked"}]

**Comment by Alys** [comment]
I'm closing this because PR #8198 added Habit counters and related features can be based on that now.
